### PR TITLE
Fragments of Data Cleaning Chapter

### PR DESCRIPTION
**Pull request by tillahoffmann**
This PR adds the first few sections of the Data Cleaning chapter; it replaces #7. Specifically,

- `01-intro.md` containing a brief overview and learning objectives.
- `02-reshaping-data.md` discussing long and wide data and transposition.
- `03-tidying-data.md` discussing row and column filtering, summary statistics, grouping, sorting, etc.
- `Makefile` that builds html versions of the markdown documents including execution. We won't need this to present the course, but it's a nice sanity check that the code in the notebooks actually runs.
- `requirements.in` and `requirements.txt` which specified the versions of various python packages to avoid runtime environment drift.
- `.github/workflows/python-markdown.yml` which sets up a python environment and executes the notebooks in CI. I may need a few iterations to get it to build; sorry about the inbox noise.

A few notes:

- I've stuck the files in a folder for the chapter to keep things organized and used `.md` extensions because the content isn't R. I don't know if that messes with the ottr build?
- Thought I'd get your input on the first few sections of the chapter before committing to the remaining sections. All input welcome!
- I've omitted the "What is Tidy Data?" section because it's not Python-specific and already well covered in the existing course.

### CODE REVIEW

**Comment by tillahoffmann** [comment]
Converted to draft for now. Will rebase and reopen once #13 is merged.

**Review by cansavvy — COMMENTED** [review]
Once you get the branches sorted out, I think the content that is here so far looks great! Very excited for this. Only some minor comments, but the general strategies and content look great to me!

File: 05-cleaning-data_02-reshaping-data.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+### Reshaping data
+
+Converting your data from wide-to-long or from long-to-wide data formats is referred to as **reshaping** your data. You can do this with the [`pandas``](https://pandas.pydata.org) package. In fact, we will use the pandas package for all of our data tidying, wrangling, and reshaping. As with most helpful Python packages, there is more functionality than what is discussed here, so feel free to explore the additional resources at the bottom to learn even more.
+
+For these examples, we'll work with the [`relig_income`](https://tidyr.tidyverse.org/reference/relig_income.html)  dataset available in `tidyr`. The data in this dataset includes the income ranges and religions from the Pew religion and income survey. The following code downloads the data from GitHub and displays the first few rows using the [`head`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.head.html) method of the dataframe.
```
[the remark this answers is not in the document]
Comment (tillahoffmann) [reply 05-cleaning-data_02-reshaping-data.md:61]: Good point. I took a somewhat content-blind approach translating the existing R resources to Python. But maybe this is a good opportunity to review some of the material, too. I'll have a look for other datasets.

File: 05-cleaning-data_02-reshaping-data.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+| \$10-20k  | 34       | 27      | 21       | 617      | 14                 | 869              | 9     | 244                     | 27                | 19     | 495           | 40     | 7      | 17       | 7               | 33           | 2                     | 299          |
+| \$20-30k  | 60       | 37      | 30       | 732      | 15                 | 1064             | 7     | 236                     | 24                | 25     | 619           | 48     | 9      | 23       | 11              | 40           | 3                     | 374          |
+| \$30-40k  | 81       | 52      | 34       | 670      | 11                 | 982              | 9     | 238                     | 24                | 25     | 655           | 51     | 10     | 32       | 13              | 46           | 4                     | 365          |
+
+While `relig_income` data had 18 rows and 11 columns, `transposed` has 11 rows and 18 columns. The information in both `relig_income` and `transposed` is the same, it's just in a different shape.
+
+
+## Additional Resources
+
+* [TidyDataTutor](https://tidydatatutor.com/vis.html).
```
[the remark this answers is not in the document]
Comment (tillahoffmann) [reply 05-cleaning-data_02-reshaping-data.md:151]:
```suggestion
* [PandasTidyDataTutor](https://pandastutor.com/)
```

**Review by tillahoffmann — COMMENTED** [review]
Thank you for the review! A few follow-up questions inline.

File: 05-cleaning-data_02-reshaping-data.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tillahoffmann) [reply 05-cleaning-data_02-reshaping-data.md:61]: I've been using the `palmerpenguins` dataset (https://allisonhorst.github.io/palmerpenguins/) for examples. It's quite neat and could probably illustrate the same concepts.

File: 05-cleaning-data_02-reshaping-data.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+This is a dataset we've looked at in a previous lesson. As discussed previously, it's a rectangular and tidy dataset. Now, we can also state that it is a wide dataset. Here you can clearly see what measurements were taken for each individual and can get a sense of how many individuals are contained in the dataset.
+
+Specifically, each individual is in a different row with each variable in a different column. At a glance we can quickly see that we have information about four different people and that each person was measured in four different ways.
+
+### Long Data
+
+Long data, on the other hand, has one column indicating the type of variable contained in that row and then a separate row for the value for that variable. Each row contains a single observation for a single variable.  Below is an example of a long data set.
+
+| ID | Variable | Value |
```
[the remark this answers is not in the document]
Comment (cansavvy) [reply 05-cleaning-data_02-reshaping-data.md:26]: The docker image OTTR uses has python on it so we could easily add to the github action to rerun the python code and then you could have it so the OTTR markdown just displays that through an iframe or something by knitting in a smaller markdown.

But this is probably not 100% necessary unless we forsee a change in the table that would change the interpretation of the chapter.

File: 05-cleaning-data_02-reshaping-data.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (cansavvy) [reply 05-cleaning-data_02-reshaping-data.md:61]: Penguins are lovely! And that's a nice dataset to work with. Definitely a classic.

[the remark this answers is not in the document]
Comment (tillahoffmann) [reply 05-cleaning-data_02-reshaping-data.md:61]: I've been pondering on this a bit more, and the penguin dataset is already "tidy" so a number of sections in the chapter aren't that relevant. Do you know of another dataset that has untidy structure but is maybe a less sensitive dataset?

**Review by davonperson — APPROVED** [review]

File: 05-cleaning-data_02-reshaping-data.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (tillahoffmann) [reply 05-cleaning-data_02-reshaping-data.md:26]: I'm prone to falling into the [automation trap](https://xkcd.com/1319/) so stuck with the copy-paste for now. Definitely up for revisiting if we end up wanting to include a lot of outputs.

File: 05-cleaning-data_02-reshaping-data.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tillahoffmann) [reply 05-cleaning-data_02-reshaping-data.md]: The `billboard` charts from 2000 seem to do the trick.

**Comment by tillahoffmann** [comment]
Thanks for the review, @davonperson! 

I have updated the example dataset to use billboard charts rather than the survey on religion and income. I had to make a few changes to the dictionary and renamed `LICENSE.md` to `LICENSE.txt` so it wouldn't get picked up by the spellchecker. I think this is now ready to be merged.

**Pull request merged**